src/app/SafeAttributePersistenceProvider.h

```diff
@@ -0,0 +1,223 @@
+#include <cstring>
+#include <inttypes.h>
+#include <lib/support/BufferReader.h>
+#include <lib/support/BufferWriter.h>
+#include <lib/support/Span.h>
+
+namespace chip {
+namespace app {
+
+/**
+ * Interface for persisting attribute values. This will always write attributes in storage as little-endian
+ * and uses a different key space from AttributePersistenceProvider.
+ */
+
+class SafeAttributePersistenceProvider
+{
+public:
+    virtual ~SafeAttributePersistenceProvider() = default;
+    SafeAttributePersistenceProvider()          = default;
+
+    // The following API provides helper functions to simplify the access of commonly used types.
+    // The API may not be complete.
+    // Currently implemented write and read types are: uint8_t, uint16_t, uint32_t, unit64_t and
+    // their nullable varieties, and bool.
+
+    /**
+     * Write an attribute value of type intX, uintX or bool to non-volatile memory.
+     *
+     * @param [in] aPath the attribute path for the data being written.
+     * @param [in] aValue the data to write.
+     */
+    template <typename T, std::enable_if_t<std::is_integral<T>::value, bool> = true>
+    CHIP_ERROR WriteScalarValue(const ConcreteAttributePath & aPath, T & aValue)
+    {
+        uint8_t value[sizeof(T)];
+        auto w = Encoding::LittleEndian::BufferWriter(value, sizeof(T));
+        w.EndianPut(uint64_t(aValue), sizeof(T));
+
+        return SafeWriteValue(aPath, ByteSpan(value));
+    }
+
+    /**
+     * Read an attribute of type intX, uintX or bool from non-volatile memory.
+     *
+     * @param [in]     aPath the attribute path for the data being persisted.
+     * @param [in,out] aValue where to place the data.
+     */
+    template <typename T, std::enable_if_t<std::is_integral<T>::value, bool> = true>
+    CHIP_ERROR ReadScalarValue(const ConcreteAttributePath & aPath, T & aValue)
+    {
+        uint8_t attrData[sizeof(T)];
+        MutableByteSpan tempVal(attrData);
+        auto err = SafeReadValue(aPath, tempVal);
+        if (err != CHIP_NO_ERROR)
+        {
+            return err;
+        }
+
+        chip::Encoding::LittleEndian::Reader r(tempVal.data(), tempVal.size());
+        r.RawReadLowLevelBeCareful(&aValue);
+        return r.StatusCode();
+    }
+
+    /**
+     * Write an attribute value of type nullable intX, uintX to non-volatile memory.
+     *
+     * @param [in] aPath the attribute path for the data being written.
+     * @param [in] aValue the data to write.
+     */
+    template <typename T, std::enable_if_t<std::is_integral<T>::value && !std::is_same<bool, T>::value, bool> = true>
+    CHIP_ERROR WriteScalarValue(const ConcreteAttributePath & aPath, DataModel::Nullable<T> & aValue)
+    {
+        if (aValue.IsNull())
+        {
+            T nullVal;
+            NumericAttributeTraits<T>::SetNull(nullVal);
+            return WriteScalarValue(aPath, nullVal);
+        }
+        return WriteScalarValue(aPath, aValue.Value());
```

Comment:
```suggestion
        NumericAttributeTraits<T>::StorageType storageValue;
        if (aValue.IsNull())
        {
            NumericAttributeTraits<T>::SetNull(storageValue);
        }
        else 
        {
            NumericAttributeTraits<T>::WorkingToStorage(aValue.Value(), storageValue);
        }
        return WriteScalarValue(aPath, storageValue);
```
and then you don't need the bool specialization.